ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -357,24 +370,58 @@ public boolean hasSucceeded() {
     return hasSucceeded;
   }
 
+  @Override
+  public boolean maybeFailedDueToOfflineReplicas() {
+    if(!routerConfig.routerUnavailableDueToOfflineReplicas) {
+      return false;
+    }
+    // We mark the failure as due to offline replicas when we know that we couldn't find the blob in eligible replicas,
+    // and remaining replicas are enough to make the request successful, and there is atleast one offline replica in the
+    // remaining set of replicas. The offline replicas can come back up in future to make the request successful, and
+    // hence such a failure should be deemed as retryable.
+    if (hasFailedOnOriginatingDcNotFound()
```

Comment:
I don't understand the logic here, Let's say we are dealing with originating dc, we have 3 replicas and 1 is offline, the other two returns NOT_FOUND and SUCCESS. It's basically what happened in the TTL_UPDATE getting 404. In this case, we want to return AmbryUnavailable so this method should return true. But, in this case, `hasFailedOnOriginatingDcNotFound` would return false. so we are not going to return true?

Reply:
In your case, `hasFailedOnOriginatingDcNotFound` will return false, but logic will try for cross-colo replicas as well, because for ttl-updates and deletes, cross-colo is enabled. And then it will eventually hit the second case for cross-colo not found.

Reply:
done.